docs/type.md

```diff
@@ -1,98 +1,197 @@
 
-This let-binding doesn't contain any written type:
+_Everything_ in Reason has a type, even if you do not write it down.
 
-<!--DOCUSAURUS_CODE_TABS-->
-<!--Reason-->
 ```reason
-let score = 10;
+let count = 42;
 ```
-<!--Output-->
-```js
-var score = 10;
+
+Types can be explicitly added with an annotation:
+
+```reason
+let count: int = 42;
 ```
-<!--END_DOCUSAURUS_CODE_TABS-->
 
-Reason deducted that `score` is an `int`, judging by the value `10`. This is called **inference**.
+Because `count` has a type the compiler knows what we are and are not allowed
+to do with its value:
 
-But types can also be explicitly written down by choice:
+```reason
+/* Allowed: addition */
+let nextCount = count + 1;
+
+/* Error: count is not a list */
+let x = List.map(fn, count);
+```
+
+## Annotations
+
+Type annotations can appear almost anywhere. They are not often necessary due to
+Reason's type inference, but they can be helpful to confirm your own
+understanding of the program's types.
+
+`int` and `string` are annotations used throughout these examples:
 
-<!--DOCUSAURUS_CODE_TABS-->
-<!--Reason-->
 ```reason
-let score: int = 10;
+let five: int = 5;
+let nine = (five: int) + (4: int);
+let add = (x: int, y: int): int => x + y;
+let drawCircle = (~radius: int): string => "hi";
 ```
-<!--Output-->
-```js
-var score = 10;
+
+## Aliases
+
+Aliases can be defined for types. This is helpful to attach meaning to simple
+types and when working with complex types that become long to write down.
+
+```reason
+type seconds = int;
+type timeInterval = (seconds, seconds);
 ```
-<!--END_DOCUSAURUS_CODE_TABS-->
 
-You can also wrap any expression in parentheses and annotate it:
+Using the alias `seconds` it is clear how the `sleep` function works. If `int`
+were used it might not be obvious:
 
-<!--DOCUSAURUS_CODE_TABS-->
-<!--Reason-->
 ```reason
-let myInt1 = 5;
-let myInt2: int = 5;
-let myInt3 = (5: int) + (4: int);
-let add = (x: int, y: int): int => x + y;
-let drawCircle = (~radius as r: int) : string => "hi";
+let sleep = (time: seconds) => { ... }
 ```
-<!--Output-->
-```js
-var myInt1 = 5;
-var myInt2 = 5;
-var myInt3 = 9;
-function addInts(x, y) {
-  return x + y | 0;
-}
-function drawCircle(r) {
-  return "hi";
-}
+
+Type aliases are required in some cases, such as working with
+[variants](variant.md) or [records](record.md).
+
+## Type Arguments
+
+Types can have arguments. This is useful when defining structures that work
+with many types of values. Having a single `list` type with an argument that can
+be `int`, `float`, or `string`, is better than having three concrete `intList`,
+`floatList`, and `stringList` types.
+
+Arguments are prefixed with a single `'` when defining the type:
+
+```reason
+type list('item) = ...
 ```
-<!--END_DOCUSAURUS_CODE_TABS-->
 
-Note: in the last line, `(~radius as r: int)` is a labeled argument. More on this [here](function.md).
+When using this type as an annotation the argument must be filled in with a
```

Comment:
> When using this type as an annotation the argument must be filled in with a concrete type:

I might say "can be" instead of "must be", just because you can also use a `'a` in these places, which means "Figure out the type for me and give it the name `'a`" (not that we have to cover that here).